core/src/main/scala/org/http4s/internal/package.scala

```diff
@@ -124,19 +129,32 @@ package object internal {
     }
 
   // Adapted from https://github.com/typelevel/cats-effect/issues/160#issue-306054982
-  private[http4s] def fromCompletableFuture[F[_], A](fcf: F[CompletableFuture[A]])(
-      implicit F: Concurrent[F]): F[A] =
-    fcf.flatMap { cf =>
-      F.cancelable { cb =>
-        cf.handle[Unit](new BiFunction[A, Throwable, Unit] {
-          override def apply(result: A, err: Throwable): Unit = err match {
-            case null => cb(Right(result))
-            case _: CancellationException => ()
-            case ex: CompletionException if ex.getCause ne null => cb(Left(ex.getCause))
-            case ex => cb(Left(ex))
+  private[http4s] def fromCompletionStage[F[_], CF[x] <: CompletionStage[x], A](
+      fcs: F[CF[A]])(implicit F: Async[F], CS: ContextShift[F]): F[A] =
+    fcs.flatMap { cs =>
+      F.async[A] { cb =>
+          cs.handle[Unit] { (result, err) =>
+            err match {
+              case null => cb(Right(result))
+              case _: CancellationException => ()
+              case ex: CompletionException if ex.getCause ne null => cb(Left(ex.getCause))
+              case ex => cb(Left(ex))
+            }
           }
-        })
-        F.delay { cf.cancel(true); () }
-      }
+          ()
+        }
```

Comment:
From the Javadoc on `cancel`:

>  Dependent CompletableFutures that have not already completed will also complete exceptionally,

I'm not sure whether dependent futures are completed exceptionally on other exceptions.  This may be a reason to do it when we know we have a `Future`.  I don't feel strongly.

Reply:
We also could to runtime casting stuff to detect `CompletableFuture`, but it is ugly. What about requiring `Concurrent` instead of `Async`, so that potential future changes to cancellation are not breaking?

Reply:
That sounds good to me.  We can always add an async version later if the need arises.